text/0000-jsbox.md

```diff
@@ -0,0 +1,341 @@
+
+```rust
+fn create_pool(mut cx: FunctionContext) -> JsResult<JsValue> {
+    let pool = Pool::new();
+
+    // How do we return a `Pool`?
+}
+
+fn get_user(mut cx: FunctionContext) -> JsResult<JsString> {
+    let id = cx.argument::<JsNumber>(0)?.value();
+    let pool = { /* How do we get a pool? */ };
+
+    let username = pool.get_by_id(id);
+
+    Ok(cx.string(username))
+}
+```
+
+The instance of `Pool` can be returned to Javascript using a `JsBox`. The `Pool` won't be dropped until the `JsBox` is garbage collected.
+
+```rust
+fn create_pool(mut cx: FunctionContext) -> JsResult<JsBox<Pool>> {
+    let pool = Pool::new();
+
+    cx.boxed(pool)
+}
+```
+
+The `Pool` instance can later be borrowed:
+
+_Note: Through `std::ops::Deref` and the magic of auto-deref, `JsBox<T>` can be treated as `&T`._
+
+```rust
+fn get_user(mut cx: FunctionContext) -> JsResult<JsString> {
+    let pool = cx.argument::<JsBox<Pool>>(0)?;
+    let id = cx.argument::<JsNumber>(1)?.value();
+
+    let username = pool.get_by_id(id)
+
+    Ok(cx.string(username))
+}
+```
+
+Data may only be borrowed immutability. However, `RefCell` can be used to introduce interior mutability with dynamic borrow checking rules:
+
+```rust
+fn create_pool(mut cx: FunctionContext) -> JsResult<JsBox<RefCell<Pool>>> {
+    let pool = RefCell::new(Pool::new());
+
+    cx.boxed(pool)
+}
+
+fn set_pool_size(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+    let size = cx.argument::<JsNumber>(1)?.value() as u32;
+    let mut pool = cx.argument::<JsPool<RefCell<Pool>>>(0)?;
+
+    pool.borrow_mut().set_size(size);
+
+    Ok(cx.undefined())
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### `neon::types::JsBox`
+
+```rust
+impl<T: Send + 'static> JsValue for JsBox<T> {}
+
+impl<T: Send + 'static> Managed for JsBox<T> {}
+
+impl<T: Send + 'static> ValueInternal for JsBox<> {}
+
+impl<T> JsBox where T: Send + 'static {
+    /// This should only fail if the VM is in a throwing state or is
+    /// in the process of stopping.
+    pub fn new<'a, C: Context<'a>>(
+        cx: &mut C,
+        v: T,
+    ) -> NeonResult<JsBox>;
+}
+```
+
+`JsBox` can be passed back and forth between Javascript and Rust like any other Js value.
+
+
+#### `std::ops::Deref`
+
+`JsBox<T>` acts as a smart-pointer and can be coerced to a `&T` by auto-deref.
+
+```rust
+impl<'a, T: Send + 'static> Deref for JsBox<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target;
+}
+```
+
+It is not possible to statically prove that two `JsBox<T>` do not alias the same pointer. For example, if a user passes the same value as two arguments to a function. Therefore, it is only safe for `JsBox<T>` to implement `Deref` and not `DerefMut`.
+
+#### `Send + 'static`
+
+`JsBox` are passed across threads and therefore must be `Send`. It is *not* necessary for `JsBox` to be `Sync` because they may only be borrowed on the main thread.
```

Comment:
Why are `JsBox` passed across threads? I couldn't figure this out from the RFC. For those who aren't fluent in Rust concurrency (I'm, like, intermediate level at best), could you explain a little more about the difference between `Send` and `Sync` and how that applies here?

Reply:
I'll reword this to emphasize that it's less critical that it's passed across threads and more relevant that ownership is moved.

Most owned types are `Send`. An exception being `Rc`.  The `Rc` docs actually do a pretty good job of explaining what `Send` is and why `Rc` *doesn't* implement it. https://doc.rust-lang.org/stable/std/rc/index.html

The `Send` protects against putting things in a `JsBox` that should leave the current context, even if they implement clone.